### PR TITLE
[MOD-13911] Add GC logic for NumericRangeTree

### PR DESCRIPTION
**Pull request by LukeMathWalker**
## Describe the changes in the pull request

Prepare `NumericRangeTree` for GC integration. 
This PR adds logic to:
- Scan a range to determine the GC delta (to be invoked in the child process of fork GC)
- Apply the GC delta to a given node (to be invoked in the parent process of fork GC)

The GC relies on the index of a node inside the slab as the node identifier. This differs from the existing C implementation, which relies on the memory address of the node as its identity (i.e. it passes a pointer over from the child to the parent GC process).

Stacked on top of #8292.

#### Which additional issues this PR fixes
1. MOD-...
2. #...

#### Main objects this PR modified
1. ...

#### Mark if applicable

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

#### Release Notes

- [ ] This PR requires release notes
- [x] This PR does not require release notes

If a release note is required (bug fix / new feature / enhancement), describe the **user impact** of this PR in the title.  

---

> [!NOTE]
> **Medium Risk**
> Touches core GC and tree-mutation paths (stats accounting, node removal, slab compaction, pointer remapping), which could impact correctness or memory usage if edge cases are missed, though it is heavily covered by new tests.
> 
> **Overview**
> **Adds garbage-collection support for `NumericRangeTree`.** The tree can now accept per-node GC deltas (`NodeGcDelta`) and apply them via `apply_gc_to_node`, updating entry/memory stats, tracking newly-empty leaves, and repairing HLL cardinality (including post-fork writes and ignored-last-block scenarios).
> 
> **Introduces maintenance operations to reclaim memory**: `trim_empty_leaves` prunes empty subtrees (freeing internal retained ranges when applicable) and `compact_if_sparse` compacts the node slab when many leaves are empty, remapping node indices safely.
> 
> Includes related robustness fixes: `GcApplyInfo` is now `Copy/Clone/Default`, `InvertedIndex::apply_gc` correctly sets `ignored_last_block` even when no last-block delta exists, numeric index exposes `scan_gc`/`apply_gc` helpers, and extensive integration/property tests were added to cover GC, trimming, compaction, and last-block edge cases.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 92713bf5662706cfe0662cc69b597c8a794b2814. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/redisearch_rs/numeric_range_tree/src/index.rs [hunk 1]
```diff
@@ -112,6 +112,46 @@ impl NumericIndex {
...
+        match self {
+            NumericIndex::Uncompressed(idx) => idx.block_ref(block_idx).map(|b| b.first_block_id()),
+            NumericIndex::Compressed(idx) => idx.block_ref(block_idx).map(|b| b.first_block_id()),
+        }
+    }
+
+    /// Apply garbage collection deltas to this index.
+    ///
+    /// Consumes the `delta` and returns information about what changed.
+    pub fn apply_gc(&mut self, delta: inverted_index::GcScanDelta) -> inverted_index::GcApplyInfo {
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/index.rs:129]: Is there a specific reason not to put `apply_gc` and `scan_gc` in gc.rs?

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+pub struct SingleNodeGcResult {
+    /// Number of entries removed from this node.
+    pub entries_removed: usize,
+    /// Bytes freed from this node.
+    pub bytes_freed: usize,
+    /// Bytes allocated (for new compacted blocks) in this node.
+    pub bytes_allocated: usize,
+    /// Whether the last block in this node index was ignored.
+    pub ignored_last_block: bool,
+    /// Whether this node became empty after GC.
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: can't we use II_GCScanStats for the first 4 members?

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+        let is_leaf = node.is_leaf();
+
+        let Some(range) = node.range_mut() else {
+            return SingleNodeGcResult::default();
+        };
+        let was_empty = range.entries().num_docs() == 0;
+
+        // Compute blocks added since fork.
+        let blocks_before = range.entries().num_blocks();
+        let last_block_idx = delta.delta.last_block_idx();
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: `blocks_before` here is a bit confusing because it reflect changes that were done after forking.
consider `current`/`parent`
and `last_block_idx` -> `last_block_idx_before_fork`

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+        let blocks_since_fork = blocks_before.saturating_sub(last_block_idx + 1);
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: The GC invariant is that the parent can only add blocks, so blocks_before >= last_block_idx + 1 should always hold. Using  saturating_sub silently masks violations of this invariant. Consider using a debug assertion and an explicit subtraction.

File: src/redisearch_rs/numeric_range_tree/src/range.rs [hunk 5]
```diff
@@ -190,6 +190,58 @@ impl NumericRange {
...
+    /// This sets the HLL registers from GC scan results and re-adds entries
+    /// from blocks that were added since the fork.
+    ///
+    /// # Arguments
+    ///
+    /// * `ignored_last_block` - Whether the last block was ignored during GC scan (from `GcApplyInfo`)
+    /// * `blocks_since_fork` - Number of new blocks added since the fork
+    /// * `registers_with_last_block` - HLL registers including the last block's cardinality
+    /// * `registers_without_last_block` - HLL registers excluding the last block's cardinality
+    pub(crate) fn reset_cardinality_after_gc(
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/range.rs:205]: should it be moved to tree/gc.rs?

File: src/redisearch_rs/numeric_range_tree/src/range.rs [hunk 6]
```diff
@@ -190,6 +190,58 @@ impl NumericRange {
...
+        } else {
+            self.hll.set_registers(*registers_with_last_block);
+            if blocks_to_rescan == 0 {
+                return; // No new blocks since fork, we're done
+            }
+        }
+
+        // Get the starting point for HLL update - iterate entries added since fork
+        let num_blocks = self.entries.num_blocks();
+        if blocks_to_rescan > num_blocks {
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/range.rs]: Can you explain when is this possible (and add docs)

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+            self.stats.empty_leaves = self.stats.empty_leaves.saturating_add(1);
+        }
+
+        // Update tree-level stats.
+        // We only update num_entries for leaves to avoid double-counting:
+        // when max_depth_range > 0, internal nodes retain ranges with duplicate
+        // entries, so decrementing for every node would over-subtract.
+        if is_leaf {
+            self.stats.num_entries = self.stats.num_entries.saturating_sub(info.entries_removed);
+        }
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs:123]: seems that the C code subtracts  entries_removed for all nodes with ranges. Is this a bug fix? If so we should open a ticket to backport the fix (since this pr is not going to be backported)
consider adding a test if it's not already covered

Also is there an alternative to `saturating_sub` that won't mask an underflow?

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+        if info.bytes_freed > info.bytes_allocated {
+            self.stats.inverted_indexes_size = self
+                .stats
+                .inverted_indexes_size
+                .saturating_sub(info.bytes_freed - info.bytes_allocated);
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: Is there a scenario where bytes_freed - bytes_allocated > inverted_indexes_size? If not, consider using a debug assertion instead of silently saturating.
(same fo the `else` block)

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+
+    /// Recursively remove empty children from a node.
+    /// Returns true if this node is empty, false otherwise.
+    fn remove_empty_children(
+        nodes: &mut NodeArena,
+        node_idx: NodeIndex,
+        rv: &mut AddResult,
+    ) -> bool {
+        let Some((left_idx, right_idx)) = nodes[node_idx].child_indices() else {
+            // Leaf node — empty iff no docs.
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]:
```suggestion
            // Leaf node — empty if no docs.
```

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+    /// Free an entire subtree, removing all slots from the slab.
+    fn free_subtree(nodes: &mut NodeArena, node_idx: NodeIndex, rv: &mut AddResult) {
+        // Read child indices before removing, to avoid borrow issues.
+        let children = nodes[node_idx].child_indices();
+
+        match &nodes[node_idx] {
+            NumericRangeNode::Leaf(leaf) => {
+                rv.num_leaves_delta -= 1;
+                rv.size_delta -= leaf.range.memory_usage() as i64;
+                rv.num_records_delta -= leaf.range.num_entries() as i32;
```
Comment (meiravgri) [inline src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: is `rv.num_records_delta` used by the caller?

File: src/redisearch_rs/numeric_range_tree/src/index.rs (hunk 1 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/index.rs:129]: I kept them in `index` since they just forward to the underlying methods with the same name. I've tried to avoid littering the rest of the crate with `match` statements over the compressed/uncompressed variants.

File: src/redisearch_rs/numeric_range_tree/src/range.rs (hunk 5 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/range.rs:205]: All `NumericRange` fields are private, so they aren't visible outside this module; that's why this is in `range.rs` rather than in `gc.rs`.

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 3 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: Tried to rename for clarity, let me know if it's better.

File: src/redisearch_rs/numeric_range_tree/src/range.rs (hunk 6 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/range.rs]: It shouldn't be, documented the rationale in a comment.

**Comment by LukeMathWalker** [comment]
I've switched all stats operations to use `checked_*` operations @meiravgri, which will panic on underflow/overflow.

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 10 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: It isn't used on the GC path.
We could remove it, by creating a distinct stat-tracking output type for `trim_empty_leaves` rather than reusing `AddResult`.

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 7 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs:123]: Looking through the C code, it does seem to be a bug fix.

Entries are counted once when added (https://github.com/RediSearch/RediSearch/blob/95f87e734de8a0114eb065cdb2d54d067bcff590/src/numeric_index.c#L366), even if they exist in both a leaf node and an internal node. 
On the removal side, we decrement the counter every time we apply a GC delta to a node (https://github.com/RediSearch/RediSearch/blob/95f87e734de8a0114eb065cdb2d54d067bcff590/src/fork_gc.c#L683), no matter if that node was a leaf or internal. So we do double-decrement.

> consider adding a test if it's not already covered

Here as a unit test or in the backport PR?

Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs:123]: Tracking it as MOD-14087.

File: src/redisearch_rs/numeric_range_tree/src/index.rs (hunk 1 above)
Comment (meiravgri) [reply src/redisearch_rs/numeric_range_tree/src/index.rs:129]: In practice we could have moved this into tree/gc.rs right?
```
impl NumericIndex {
    // apply_gc
   // scan_gc
}
``

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 10 above)
Comment (meiravgri) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: possible

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 7 above)
Comment (meiravgri) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs:123]: Perfect thanks :)

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 3 above)
Comment (meiravgri) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: Very better 😄

File: src/redisearch_rs/numeric_range_tree/src/lib.rs [hunk 11]
```diff
@@ -82,5 +82,5 @@ pub use inverted_index::NumericFilter;
 pub use iter::ReversePreOrderDfsIterator;
 pub use node::{InternalNode, LeafNode, NumericRangeNode};
 pub use range::NumericRange;
-pub use tree::{AddResult, NumericRangeTree};
+pub use tree::{AddResult, NodeGcDelta, NumericRangeTree, SingleNodeGcResult};
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/lib.rs]: Fixed in https://github.com/RediSearch/RediSearch/pull/8293/changes/54d2f47ad1a1d69feb607b7074c13b0df9335a2d (cc @meiravgri )

File: src/redisearch_rs/numeric_range_tree/src/tree/gc.rs (hunk 10 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/tree/gc.rs]: Done in https://github.com/RediSearch/RediSearch/pull/8293/changes/54d2f47ad1a1d69feb607b7074c13b0df9335a2d

File: src/redisearch_rs/numeric_range_tree/src/index.rs (hunk 1 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/index.rs:129]: Yes, that would be possible since both enum variants are public.

File: src/redisearch_rs/numeric_range_tree/src/range.rs [hunk 12]
```diff
@@ -190,6 +190,60 @@ impl NumericRange {
...
+        let mut blocks_to_rescan = blocks_since_fork;
+
+        if ignored_last_block {
+            self.hll.set_registers(*registers_without_last_block);
+            blocks_to_rescan += 1; // The last block was ignored, so re-add it too
+        } else {
+            self.hll.set_registers(*registers_with_last_block);
+            if blocks_to_rescan == 0 {
+                return; // No new blocks since fork, we're done
+            }
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/redisearch_rs/numeric_range_tree/src/range.rs:221]: Fixed in https://github.com/RediSearch/RediSearch/commit/54d2f47ad1a1d69feb607b7074c13b0df9335a2d (cc @meiravgri )

**Review by meiravgri — APPROVED** [review]
